perun-core/src/main/java/cz/metacentrum/perun/core/impl/ExtSourceTCS.java

```diff
@@ -0,0 +1,303 @@
+						String extractedCert = exportBase64Certificate(pemObject);
+						//For this user, add his valid certificate to the map
+						validCertificatesForLogin.put(login, new Pair<>(certificateHolder, extractedCert));
+					} else {
+						//there is already a certificate for this user, you need to compare them on expiration date
+						Date dayOfExpirationOfTheNewCertificate = certificateHolder.getNotAfter();
+						Date dayOfExpirationOfSavedCertificate = validCertificatesForLogin.get(login).getLeft().getNotAfter();
+						//If certificate saved in the structure expires sooner than this certificate, replace it with this one
+						if (dayOfExpirationOfTheNewCertificate.after(dayOfExpirationOfSavedCertificate)) {
+							String extractedCert = exportBase64Certificate(pemObject);
+							validCertificatesForLogin.put(login, new Pair<>(certificateHolder, extractedCert));
+						}
+					}
+					//read another certificate (or null if this was the last one)
+					pemObject = pemParser.readPemObject();
+				}
+			}
+		} catch (IOException ex) {
+			throw new InternalErrorException(ex);
+		} finally {
+			if(con != null) {
+				con.disconnect();
+			}
+		}
+
+		return validCertificatesForLogin;
+	}
+
+	/**
+	 * Check if x509Certificate is valid and return login parsed from subject of such valid certificate.
+	 * Throw an exception if certificate is not valid or check can't be done correctly.
+	 *
+	 * Valid certificate means:
+	 *  - it is not expired
+	 *  - login of the owner can be parsed from certificate subject
+	 *  - every valid owner already exists in perun as User
+	 *
+	 * @param x509CertificateHolder certificate in x509Holder object
+	 * @return if certificate is valid, return login of owner from it's subject
+	 * @throws InternalErrorException if there is any problem with getting all logins from Perun (to check existence of owner in Perun)
+	 * @throws InvalidCertificateException if certificate is not valid, throw an exception
+	 */
+	private String checkCertAndGetLogin(X509CertificateHolder x509CertificateHolder) throws InternalErrorException, InvalidCertificateException {
+		Date now = new Date();
+		//skip expired certificates
+		Date dayOfCertificateExpiration = x509CertificateHolder.getNotAfter();
+		if (dayOfCertificateExpiration.before(now)) throw new InvalidCertificateException("Certificate is already expired.");
+
+		//skip wrong certificates and parse UCO from the subject
+		String subject = x509CertificateHolder.getSubject().toString();
+		Matcher loginMatcher = loginPattern.matcher(subject);
+		Matcher wrongLoginMatcher = wrongLoginPattern.matcher(subject);
+		if (wrongLoginMatcher.matches()) throw new InvalidCertificateException("There is more than one login in certificate's subject.");
+		if (!loginMatcher.matches()) throw new InvalidCertificateException("There is missing login in certificate's subject.");
+
+		String login = loginMatcher.group(1);
+
+		List<String> allLogins = getLoginsFromPerun(attrLoginMUName);
```

Comment:
We could probably get all logins from perun once and not for each certificate.

Reply:
Great idea! Done.